### PR TITLE
add Hermitian support for shot-noise & LAPACK linking

### PR DESCRIPTION
**Pull request by multiphaseCFD**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**

Add `Hermitian` observable support for stochastic measurement to C++ layer. `LAPACK` is required to solve `eigenvalues` and the `unitary` of a given `Hermitian`. The main challenge of this task is `LAPACK` linking for Windows system. To solve this issue, `vcpkg` is used to install `LAPACK` and the only thing Windows users need to do is to specify the path to set `CMAKE_TOOLCHAIN_FILE` with `-DCMAKE_TOOLCHAIN_FILE=<vcpkg-root>/scripts/buildsystems/vcpkg.cmake`.

**Description of the Change:**

C++ backends now can link against LAPACK.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by trbromley** [comment]
[sc-51606]

File: pennylane_lightning/core/src/measurements/tests/Test_MeasurementsBase.cpp [hunk 1]
```diff
@@ -580,25 +585,83 @@ template <typename TypeList> void testHermitianObsExpvalShot() {
 
         // Initializing the measures class.
```
Comment (AmintorDusko) [inline pennylane_lightning/core/src/measurements/tests/Test_MeasurementsBase.cpp:586]:
```suggestion
        // Initializing the Measurements class.
```

File: pennylane_lightning/core/src/measurements/tests/Test_MeasurementsBase.cpp [hunk 2]
```diff
@@ -694,6 +758,9 @@ template <typename TypeList> void testTensorProdObsExpvalShot() {
         using PrecisionT = typename StateVectorT::PrecisionT;
         using ComplexT = StateVectorT::ComplexT;
 
+#ifdef PL_USE_LAPACK
+        using MatrixT = std::vector<ComplexT>;
+#endif
```
Comment (AmintorDusko) [inline pennylane_lightning/core/src/measurements/tests/Test_MeasurementsBase.cpp:763]: Apparently, this typedef definition is being used only once, in the block starting at line 845.
In this case, can we have this definition moved there?

**Review by AmintorDusko — COMMENTED** [review]
Comments and suggestions, so far.

File: README.rst [hunk 3]
```diff
@@ -389,6 +389,24 @@ where ``${TARGET}`` is one of the following
 
 .. docker-end-inclusion-marker-do-not-remove
 
+LAPACK support
+**************
+
+LAPACK, a numerical linear algebra library, is requiredd to enable stochastic measurement support in C++ backend. We suggest first install ``LAPACK`` following instrcutions in `LAPACK document <https://github.com/Reference-LAPACK/lapack/blob/db501d91ba690192dfe946524251b02d416298d0/README.md?plain=1#L85-L92>`_.
```
[the remark this answers is not in the document]
Comment (multiphaseCFD) [reply README.rst:395]: Nice spot! Thanks

File: README.rst [hunk 4]
```diff
@@ -389,6 +389,24 @@ where ``${TARGET}`` is one of the following
...
+.. code-block:: console
+
+    $ sudo apt -y update &&
+    $ sudo apt install g++ liblapack-dev
+
+where ``liblapack-dev`` is included to also install LAPACK.
+
+On Windows systems, LAPACK is recommended to be built and installed using `vcpkg <https://github.com/Microsoft/vcpkg/>`_ following the instructions `here <https://github.com/Reference-LAPACK/lapack/blob/db501d91ba690192dfe946524251b02d416298d0/README.md?plain=1#L85-L92>`_.
+
+The CMake option ``-DENABLE_LAPACK=ON`` must also be specified when building C++ backends. For Windows systems, we suggest to add the CMake option ``-DCMAKE_TOOLCHAIN_FILE=<vcpkg-root>/scripts/buildsystems/vcpkg.cmake`` to ensure ``LAPACK`` can be found with cmake.
```
[the remark this answers is not in the document]
Comment (multiphaseCFD) [reply README.rst:408]: Done!

File: pennylane_lightning/core/src/measurements/tests/Test_MeasurementsBase.cpp (hunk 1 above)
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/measurements/tests/Test_MeasurementsBase.cpp:586]: Done!

File: pennylane_lightning/core/src/measurements/tests/Test_MeasurementsBase.cpp (hunk 2 above)
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/measurements/tests/Test_MeasurementsBase.cpp:763]: Let me move it.

File: .github/workflows/tests_windows.yml [hunk 5]
```diff
@@ -22,7 +22,7 @@ concurrency:
 jobs:
   cpptests:
     if: ${{ !contains(fromJSON('["schedule", "workflow_dispatch"]'), github.event_name) }}
-    timeout-minutes: 30
+    timeout-minutes: 120
```
[the remark this answers is not in the document]
Comment (multiphaseCFD) [reply .github/workflows/tests_windows.yml:25]: No really, it's not a temporary change since only the lapack installation takes about 20 minutes.

File: .github/workflows/tests_gpu_kokkos.yml [hunk 6]
```diff
@@ -86,14 +86,12 @@ jobs:
         env:
           CMAKE_PREFIX_PATH: ${{ github.workspace }}/Kokkos_install/${{ matrix.exec_model }}
         run: |
-          set -x
```
[the remark this answers is not in the document]
Comment (multiphaseCFD) [reply .github/workflows/tests_gpu_kokkos.yml:89]: yea, it enables debug mode. we can revert it.

File: .github/workflows/tests_linux.yml [hunk 7]
```diff
@@ -47,13 +47,14 @@ jobs:
...
       - name: Install dependencies
-        run: echo ${{ github.event_name }} && sudo apt-get update && sudo apt-get -y -q install cmake gcc-$GCC_VERSION  g++-$GCC_VERSION  ninja-build gcovr lcov
+        run: echo ${{ github.event_name }} && sudo apt-get update && sudo apt-get -y -q install cmake gcc-$GCC_VERSION  g++-$GCC_VERSION liblapack-dev ninja-build gcovr lcov
 
       - name: Build and run unit tests
         run: |
             cmake . -BBuild -G Ninja \
               -DCMAKE_BUILD_TYPE=Debug \
               -DBUILD_TESTS=ON \
+              -DENABLE_LAPACK=ON \
```
Comment (maliasadi) [inline .github/workflows/tests_linux.yml:57]: do we need to test Lightning with `ENABLE_LAPACK=OFF` to ensure the support in both cases?

File: .github/workflows/tests_linux.yml [hunk 8]
```diff
@@ -207,6 +208,47 @@ jobs:
...
+              -DCMAKE_CXX_COMPILER=$(which g++-$GCC_VERSION) \
+              -DENABLE_COVERAGE=ON
+            cmake --build ./Build
+            cd ./Build
+            mkdir -p ./tests/results
+            for file in *runner ; do ./$file --order lex --reporter junit --out ./tests/results/report_$file.xml; done;
+            lcov --directory . -b ../pennylane_lightning/core/src --capture --output-file coverage.info
+            lcov --remove coverage.info '/usr/*' --output-file coverage.info
+            mv coverage.info coverage-${{ github.job }}-${{ matrix.pl_backend }}.info
```
Comment (maliasadi) [inline .github/workflows/tests_linux.yml:220]: why are these coverage commands added here?

File: .github/workflows/tests_linux.yml [hunk 9]
```diff
@@ -207,6 +208,47 @@ jobs:
...
+        run: sudo apt-get update && sudo apt-get -y -q install cmake gcc-$GCC_VERSION  g++-$GCC_VERSION liblapack-dev libopenblas-dev ninja-build gcovr lcov
+
+      - name: Build and run unit tests
+        run: |
+            cmake . -BBuild -G Ninja \
+              -DCMAKE_BUILD_TYPE=Debug \
+              -DENABLE_PYTHON=OFF \
+              -DPL_BACKEND=${{ matrix.pl_backend }} \
+              -DBUILD_TESTS=ON \
+              -DENABLE_LAPACK=ON \
```
Comment (maliasadi) [inline .github/workflows/tests_linux.yml:251]: to imply with the name of the job: `cpptestswithOpenBLASNoLAPACK`?
```suggestion
              -DENABLE_LAPACK=OFF \
```

File: .github/workflows/tests_linux.yml [hunk 10]
```diff
@@ -630,13 +673,15 @@ jobs:
           cp -rf ${{ github.workspace}}/Kokkos_install/${{ matrix.exec_model }}/* Kokkos/
 
       - name: Install dependencies
-        run: sudo apt-get update && sudo apt-get -y -q install cmake gcc-$GCC_VERSION  g++-$GCC_VERSION  ninja-build gcovr lcov
+        run: sudo apt-get update && sudo apt-get -y -q install cmake gcc-$GCC_VERSION  g++-$GCC_VERSION liblapack-dev ninja-build gcovr lcov
```
Comment (maliasadi) [inline .github/workflows/tests_linux.yml:676]:
```suggestion
        run: sudo apt-get update && sudo apt-get -y -q install cmake gcc-$GCC_VERSION  g++-$GCC_VERSION liblapack-dev ninja-build gcovr lcov
```

File: .github/workflows/tests_linux.yml [hunk 11]
```diff
@@ -630,13 +673,15 @@ jobs:
           cp -rf ${{ github.workspace}}/Kokkos_install/${{ matrix.exec_model }}/* Kokkos/
 
       - name: Install dependencies
-        run: sudo apt-get update && sudo apt-get -y -q install cmake gcc-$GCC_VERSION  g++-$GCC_VERSION  ninja-build gcovr lcov
+        run: sudo apt-get update && sudo apt-get -y -q install cmake gcc-$GCC_VERSION  g++-$GCC_VERSION liblapack-dev ninja-build gcovr lcov 
 
       - name: Build and run unit tests
         run: |
+            set -x
```
Comment (maliasadi) [inline .github/workflows/tests_linux.yml:680]: why is this required?

File: README.rst [hunk 12]
```diff
@@ -389,6 +389,24 @@ where ``${TARGET}`` is one of the following
...
+**************
+
+LAPACK, a numerical linear algebra library, is required to enable stochastic measurement support in the C++ backend. We suggest first installing ``LAPACK`` following instructions in `LAPACK document <https://github.com/Reference-LAPACK/lapack/blob/db501d91ba690192dfe946524251b02d416298d0/README.md?plain=1#L85-L92>`_.
+
+On Debian-based Linux systems, LAPACK can be also installed via ``apt``:
+
+.. code-block:: console
+
+    $ sudo apt -y update &&
+    $ sudo apt install g++ liblapack-dev
```
Comment (maliasadi) [inline README.rst:402]: is `g++` necessary here?
 
```suggestion
    $ sudo apt install liblapack-dev
```

File: README.rst [hunk 13]
```diff
@@ -389,6 +389,24 @@ where ``${TARGET}`` is one of the following
...
+
+where ``liblapack-dev`` is included to also install LAPACK.
```
Comment (maliasadi) [inline README.rst:404]:
```suggestion
where LAPACK is included in ``liblapack-dev``.
```

File: README.rst [hunk 14]
```diff
@@ -389,6 +389,24 @@ where ``${TARGET}`` is one of the following
...
+
+On Windows systems, LAPACK is recommended to be built and installed using `vcpkg <https://github.com/Microsoft/vcpkg/>`_ following the instructions `here <https://github.com/Reference-LAPACK/lapack/blob/db501d91ba690192dfe946524251b02d416298d0/README.md?plain=1#L85-L92>`_.
```
Comment (maliasadi) [inline README.rst:406]: please add a brief description to install LAPACK on macOS.

File: pennylane_lightning/core/src/observables/Observables.hpp [hunk 15]
```diff
@@ -235,6 +246,26 @@ class HermitianObsBase : public Observable<StateVectorT> {
...
+                           return static_cast<std::complex<PrecisionT>>(value);
+                       });
+
+        std::vector<std::complex<PrecisionT>> unitary(matrix_.size());
+
+        Pennylane::Util::compute_diagonalizing_gates<PrecisionT>(
+            Util::exp2(wires_.size()), Util::exp2(wires_.size()), mat,
+            eigenVals_, unitary);
+
+        unitary_.resize(unitary.size());
```
Comment (maliasadi) [inline pennylane_lightning/core/src/observables/Observables.hpp:264]: why is this necessary here?

File: pennylane_lightning/core/src/observables/Observables.hpp [hunk 16]
```diff
@@ -255,11 +286,29 @@ class HermitianObsBase : public Observable<StateVectorT> {
...
+
+        PL_ABORT_IF_NOT(
+            Pennylane::Util::is_Hermitian<PrecisionT>(Util::exp2(wires_.size()),
+                                                      Util::exp2(wires_.size()),
+                                                      mat) == true,
+            "The matrix passed to HermitianObs is not a Hermitian matrix.");
+
+        eigenValues.clear();
+        ob_wires.clear();
+        ob_wires = wires_;
```
Comment (maliasadi) [inline pennylane_lightning/core/src/observables/Observables.hpp:305]: the clear method is not required if you do the copy-assignment in the next line. right?
```suggestion
        ob_wires = wires_;
```

File: pennylane_lightning/core/src/observables/Observables.hpp [hunk 17]
```diff
@@ -255,11 +286,29 @@ class HermitianObsBase : public Observable<StateVectorT> {
...
+        sv.applyMatrix(unitary_, wires_);
+        eigenValues.push_back(eigenVals_);
```
Comment (maliasadi) [inline pennylane_lightning/core/src/observables/Observables.hpp:307]: consider using "shirnk_to_fit" to reduce the capacity to size after this line:
```suggestion
        eigenValues.push_back(eigenVals_);
        eigenValues.shirnk_to_fit();
```
https://en.cppreference.com/w/cpp/container/vector/shrink_to_fit

File: pennylane_lightning/core/src/observables/tests/Test_Observables.cpp [hunk 18]
```diff
@@ -207,6 +207,29 @@ template <typename TypeList> void testHermitianObsBase() {
             REQUIRE(ob2 != ob3);
         }
 
+#ifdef PL_USE_LAPACK
+        DYNAMIC_SECTION("Failed to create a HermitianObs- "
+                        << StateVectorToName<StateVectorT>::name) {
+            std::mt19937_64 re{1337};
+            const size_t num_qubits = 3;
```
Comment (maliasadi) [inline pennylane_lightning/core/src/observables/tests/Test_Observables.cpp:214]: also in the other test units:
```suggestion
            constexpr size_t num_qubits = 3;
```

File: pennylane_lightning/core/src/utils/CMakeLists.txt [hunk 19]
```diff
@@ -3,6 +3,11 @@ cmake_minimum_required(VERSION 3.20)
 project(lightning_utils LANGUAGES CXX)
 
 set(LQUBIT_UTILS_FILES RuntimeInfo.cpp CACHE INTERNAL "" FORCE)
+
+if(ENABLE_LAPACK)
+    list( APPEND LQUBIT_UTILS_FILES UtilLinearAlg.cpp)
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/CMakeLists.txt:8]:
```suggestion
    list(APPEND LQUBIT_UTILS_FILES UtilLinearAlg.cpp)
```

File: pennylane_lightning/core/src/utils/Util.hpp [hunk 20]
```diff
@@ -533,4 +533,29 @@ auto kronProd(const std::vector<T> &diagA, const std::vector<T> &diagB)
...
+ * @param n Number of columns.
+ * @param lda Number of rows.
+ * @param mat A matrix to be checked.
+ *
+ * @return is_Hermitian Is the matrix a Hermitian matrix or not.
+ */
+template <typename T>
+bool is_Hermitian(size_t n, size_t lda,
+                  const std::vector<std::complex<T>> &mat) {
+    for (size_t i = 0; i < n; i++) {
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/Util.hpp:551]: you can make the check parallel using OMP parallel-for ;) not required for this PR though

File: pennylane_lightning/core/src/utils/Util.hpp [hunk 21]
```diff
@@ -533,4 +533,29 @@ auto kronProd(const std::vector<T> &diagA, const std::vector<T> &diagB)
...
+        for (size_t j = i + 1; j < lda; j++) {
+            if (mat[j + i * lda] != std::conj(mat[i + j * n])) {
+                return false;
+            }
+        }
+    }
+    return true;
+}
+
 } // namespace Pennylane::Util
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/Util.hpp:561]:
```suggestion
} // namespace Pennylane::Util

```

File: pennylane_lightning/core/src/utils/UtilLinearAlg.cpp [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+
+#include "UtilLinearAlg.hpp"
+
+template void Pennylane::Util::compute_diagonalizing_gates<float>(
+    int n, int lda, const std::vector<std::complex<float>> &Ah,
+    std::vector<float> &eigenVals, std::vector<std::complex<float>> &unitary);
+
+template void Pennylane::Util::compute_diagonalizing_gates<double>(
+    int n, int lda, const std::vector<std::complex<double>> &Ah,
+    std::vector<double> &eigenVals, std::vector<std::complex<double>> &unitary);
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/UtilLinearAlg.cpp:31]:
```suggestion
    std::vector<double> &eigenVals, std::vector<std::complex<double>> &unitary);

```

File: pennylane_lightning/core/src/utils/UtilLinearAlg.hpp [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+extern void cheev_(const char *jobz, const char *uplo, const int *n,
+                   std::complex<float> *a, const int *lda, float *w,
+                   std::complex<float> *work, const int *lwork, float *rwork,
+                   int *info);
+extern void zheev_(const char *jobz, const char *uplo, const int *n,
+                   std::complex<double> *a, const int *lda, double *w,
+                   std::complex<double> *work, const int *lwork, double *rwork,
+                   int *info);
+}
+} // namespace
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/UtilLinearAlg.hpp:24]: nice trick!

File: pennylane_lightning/core/src/utils/UtilLinearAlg.hpp [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+ */
+template <typename T>
+void compute_diagonalizing_gates(int n, int lda,
+                                 const std::vector<std::complex<T>> &Ah,
+                                 std::vector<T> &eigenVals,
+                                 std::vector<std::complex<T>> &unitary) {
+    eigenVals.clear();
+    eigenVals.resize(n);
+    unitary.clear();
+    unitary = std::vector<std::complex<T>>(n * n, {0, 0});
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/UtilLinearAlg.hpp:47]: do you need to clear the unitary matrix?
```suggestion
    unitary = std::vector<std::complex<T>>(n * n, {0, 0});
```

File: pennylane_lightning/core/src/utils/UtilLinearAlg.hpp [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+
+    std::vector<std::complex<T>> ah(n * lda, {0.0, 0.0});
+
+    for (size_t i = 0; i < static_cast<size_t>(n); i++) {
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/UtilLinearAlg.hpp:51]: this nested for-loop can also be rewritten in a more performant way. you may want to add a TODO or note to not forget about this in the future.

File: pennylane_lightning/core/src/utils/UtilLinearAlg.hpp [hunk 26]
```diff
@@ -0,0 +1,91 @@
...
+        // Perform eigenvalue and eigenvector computation
+        zheev_(&jobz, &uplo, &n, ah.data(), &lda, eigenVals.data(),
+               work_optimal.data(), &lwork, rwork.data(), &info);
+    }
+
+    std::transform(ah.begin(), ah.end(), unitary.begin(),
+                   [](std::complex<T> value) {
+                       return std::complex<T>{value.real(), -value.imag()};
+                   });
+}
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/UtilLinearAlg.hpp:90]: :1st_place_medal:

File: pennylane_lightning/core/src/utils/UtilLinearAlg.hpp [hunk 27]
```diff
@@ -0,0 +1,91 @@
...
+} // namespace Pennylane::Util
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/UtilLinearAlg.hpp:91]:
```suggestion
} // namespace Pennylane::Util

```

**Review by maliasadi — COMMENTED** [review]
Thanks @multiphaseCFD! :100:

File: .github/workflows/tests_linux.yml (hunk 9 above)
Comment (multiphaseCFD) [reply .github/workflows/tests_linux.yml:251]: Nice spot! Thanks

File: .github/workflows/tests_linux.yml (hunk 7 above)
Comment (multiphaseCFD) [reply .github/workflows/tests_linux.yml:57]: Thanks @maliasadi!It's done in the `cpptestswithOpenBLASNoLAPACK` job.

File: .github/workflows/tests_linux.yml (hunk 8 above)
Comment (multiphaseCFD) [reply .github/workflows/tests_linux.yml:220]: Actually, it's kind of GA issue. This difference should be shown in the `cpptestswithOpenBLASNoLAPACK` section.

File: .github/workflows/tests_windows.yml [hunk 28]
```diff
@@ -38,6 +38,17 @@ jobs:
...
 
+      - name: Install lapack with vcpkg
+        run: |
+          pushd .
+          cd ~
+          git clone --quiet --recursive https://github.com/microsoft/vcpkg.git
+          cd vcpkg
+          .\bootstrap-vcpkg.bat
+          popd
+          vcpkg integrate install
```
Comment (maliasadi) [inline .github/workflows/tests_windows.yml:49]: I'm not sure if this is required here. from the docs https://learn.microsoft.com/en-us/vcpkg/commands/integrate#vcpkg-integrate-install, this feature is embedded in VS17. 

```
On Windows with Visual Studio 2015, this subcommand will add redirecting logic into the MSBuild installation which will automatically pick up each user's user-wide vcpkg instance. Visual Studio 2017 and newer have this logic in the box.
```

File: .github/workflows/tests_linux.yml (hunk 11 above)
Comment (multiphaseCFD) [reply .github/workflows/tests_linux.yml:680]: We enable the debug mode here to allow us to track where does the failure come from.

File: README.rst (hunk 12 above)
Comment (multiphaseCFD) [reply README.rst:402]: Nice catch!

File: pennylane_lightning/core/src/utils/CMakeLists.txt (hunk 19 above)
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/utils/CMakeLists.txt:8]: Nice catch!

File: README.rst (hunk 14 above)
Comment (multiphaseCFD) [reply README.rst:406]: Done!

File: pennylane_lightning/core/src/utils/UtilLinearAlg.hpp (hunk 23 above)
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/utils/UtilLinearAlg.hpp:24]: Thanks!

File: .github/workflows/tests_windows.yml [hunk 29]
```diff
@@ -38,6 +38,17 @@ jobs:
...
+          vcpkg install lapack
```
Comment (multiphaseCFD) [inline .github/workflows/tests_windows.yml:50]: Let's remove it and see.

File: .github/workflows/tests_windows.yml (hunk 28 above)
Comment (multiphaseCFD) [reply .github/workflows/tests_windows.yml:49]: Let's remove it and see.

File: pennylane_lightning/core/src/observables/Observables.hpp (hunk 17 above)
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/observables/Observables.hpp:307]: Thanks! Since `eigenValues` is a `std::vector<std::vector<T>>` type, `std::vector<T>` doesn't meet the type requirement here. So I will keep current implementation.

File: pennylane_lightning/core/src/observables/Observables.hpp [hunk 30]
```diff
@@ -235,6 +246,26 @@ class HermitianObsBase : public Observable<StateVectorT> {
...
+                       });
+
+        std::vector<std::complex<PrecisionT>> unitary(matrix_.size());
+
+        Pennylane::Util::compute_diagonalizing_gates<PrecisionT>(
+            Util::exp2(wires_.size()), Util::exp2(wires_.size()), mat,
+            eigenVals_, unitary);
+
+        unitary_.resize(unitary.size());
+        std::transform(
```
Comment (multiphaseCFD) [inline pennylane_lightning/core/src/observables/Observables.hpp:265]: Not really, removed.

File: pennylane_lightning/core/src/observables/Observables.hpp (hunk 15 above)
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/observables/Observables.hpp:264]: Not really, removed.

File: pennylane_lightning/core/src/utils/UtilLinearAlg.hpp (hunk 24 above)
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/utils/UtilLinearAlg.hpp]: Done!

File: pennylane_lightning/core/src/observables/Observables.hpp (hunk 16 above)
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/observables/Observables.hpp]: Done!

File: pennylane_lightning/core/src/utils/Util.hpp (hunk 20 above)
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/utils/Util.hpp:552]: Done!

File: pennylane_lightning/core/src/utils/UtilLinearAlg.hpp (hunk 25 above)
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/utils/UtilLinearAlg.hpp:51]: Done!

**Review by AmintorDusko — COMMENTED** [review]
Hi Shuli nice work so far. Let me know when updates are done and CIs successful and I will give one more round of review.

File: pennylane_lightning/core/src/utils/UtilLinearAlg.hpp [hunk 31]
```diff
@@ -0,0 +1,91 @@
+
+/**
+ * @file
+ * LAPACK wrapper functions declaritions.
+ */
```
[the remark this answers is not in the document]
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/utils/UtilLinearAlg.hpp:5]: Thanks @AmintorDusko! Nice catch!

File: pennylane_lightning/core/src/utils/UtilLinearAlg.hpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/utils/UtilLinearAlg.hpp:24]: Nice one!!

File: pennylane_lightning/core/src/utils/tests/Test_Util.cpp [hunk 32]
```diff
@@ -153,6 +153,23 @@ TEMPLATE_TEST_CASE("Util::squaredNorm", "[Util][LinearAlgebra]", float,
     }
 }
 
+TEMPLATE_TEST_CASE("Util::is_Hermitian", "[Util][LinearAlgebra]", float,
+                   double) {
+    SECTION("Test a Hermition matrix") {
```
[the remark this answers is not in the document]
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/utils/tests/Test_Util.cpp:158]: Done!

File: pennylane_lightning/core/src/utils/tests/Test_Util.cpp [hunk 33]
```diff
@@ -153,6 +153,23 @@ TEMPLATE_TEST_CASE("Util::squaredNorm", "[Util][LinearAlgebra]", float,
...
+        std::vector<std::complex<TestType>> A{
+            {-6.0, 0.0}, {2.0, 1.0}, {2.0, -1.0}, {0.0, 0.0}};
+
+        REQUIRE(is_Hermitian(2, 2, A) == true);
+    }
+
+    SECTION("Test a matrix not a Hermitian") {
```
[the remark this answers is not in the document]
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/utils/tests/Test_Util.cpp]: Thanks @AmintorDusko ! Done

File: pennylane_lightning/core/src/utils/UtilLinearAlg.hpp [hunk 34]
```diff
@@ -0,0 +1,107 @@
...
+                                 const std::vector<std::complex<T>> &Ah,
+                                 std::vector<T> &eigenVals,
+                                 std::vector<std::complex<T>> &unitary) {
+    eigenVals.clear();
+    eigenVals.resize(n);
+    unitary = std::vector<std::complex<T>>(n * n, {0, 0});
+
+    std::vector<std::complex<T>> ah(n * lda, {0.0, 0.0});
+
+    // TODO optmize transpose
```
Comment (maliasadi) [inline pennylane_lightning/core/src/utils/UtilLinearAlg.hpp:66]: You didn't end up using `Utils::Transpose`. right?

**Review by maliasadi — APPROVED** [review]
Nice job @multiphaseCFD! :tada:

File: pennylane_lightning/core/src/utils/UtilLinearAlg.hpp (hunk 34 above)
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/utils/UtilLinearAlg.hpp:66]: Hey @maliasadi , The Transpose API now lives in the `lightning_qubit/util` . I will leave it as it is for now and will create a follow-up PR to reorganize the repo a little bit.

**Review by AmintorDusko — APPROVED** [review]
Nothing more to add. Nice work!
